### PR TITLE
feat(npm): support pnpm v9

### LINKED ISSUE #28467: Support pnpm v9 lockfiles
### Describe the proposed change(s).

pnpm v9 brings a new lockfile format `9.0` which is different from `6.0`

https://github.com/pnpm/pnpm/releases/tag/v9.0.0

### PR DESCRIPTION
**Pull request by viceice**
## Changes
- fix: don't delete lockfile in `writeExistingFiles`, they will be deleted `generateLockFile`
- fix: from v6 the lockfile version is string
- feat: add lockfile v9 support

## Context
- closes #28467
- repro: https://github.com/renovate-reproductions/28467-pnpm-v9

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 1]
```diff
@@ -172,25 +177,31 @@ export async function getConstraintFromLockFile(
...
- lowerConstriant : lowest pnpm version that can generate the lockfileVersion
- upperConstraint : highest pnpm version that can generate the lockfileVersion
- lowerBound      : highest pnpm version that is less than the lowerConstraint
- upperBound      : lowest pnpm version that is greater than upperConstraint
-
- For handling future lockfileVersions, we need to:
- 1. add a upperBound and upperConstraint to the current lastest lockfileVersion
- 2. add an object for the new lockfileVersion with lowerBound and lowerConstraint
+ * pnpm lockfiles have corresponding version numbers called "lockfileVersion"
+ * each lockfileVersion can only be generated by a certain pnpm version ranges
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/pnpm.ts:181]:
```suggestion
 * each lockfileVersion can only be generated by a certain pnpm version range
```

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 2]
```diff
@@ -172,25 +177,31 @@ export async function getConstraintFromLockFile(
...
+ * eg. lockfileVersion: 5.4 can only be generated by pnpm version >=7 && <8
+ * official list can be found here : https://github.com/pnpm/spec/tree/master/lockfile
+ * we use the mapping present below to find the compatible pnpm version range for a given lockfileVersion
+ *
+ * the various terms used in the mapping are explained below:
+ * lowerConstriant : lowest pnpm version that can generate the lockfileVersion
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/pnpm.ts:187]:
```suggestion
 * lowerConstraint : lowest pnpm version that can generate the lockfileVersion
```

I think you meant `lowerConstraint` here. Also check the spelling in the code. 😉

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 3]
```diff
@@ -172,25 +177,31 @@ export async function getConstraintFromLockFile(
...
+ * upperConstraint : highest pnpm version that can generate the lockfileVersion
+ * lowerBound      : highest pnpm version that is less than the lowerConstraint
+ * upperBound      : lowest pnpm version that is greater than upperConstraint
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/pnpm.ts:190]:
```suggestion
 * upperBound      : lowest pnpm version that is greater than the upperConstraint
```

Let's add a `the` here so the format matches the other explanations.

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 4]
```diff
@@ -172,25 +177,31 @@ export async function getConstraintFromLockFile(
...
+ *
+ * For handling future lockfileVersions, we need to:
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/pnpm.ts:192]:
```suggestion
 * To handle future lockfileVersions, we need to:
```

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 5]
```diff
@@ -172,25 +177,31 @@ export async function getConstraintFromLockFile(
...
+ * 1. add a upperBound and upperConstraint to the current lastest lockfileVersion
```
Comment (HonkingGoose) [inline lib/modules/manager/npm/post-update/pnpm.ts:193]:
```suggestion
 * 1. add a upperBound and upperConstraint to the current latest lockfileVersion
```

Fixing a typo.

**Review by HonkingGoose — COMMENTED** [review]
Some comment comments. 🙃

File: lib/modules/manager/npm/post-update/index.ts [hunk 6]
```diff
@@ -226,13 +226,6 @@ export async function writeExistingFiles(
     if (yarnLock && config.reuseLockFiles === false) {
       await deleteLocalFile(yarnLock);
     }
-    // istanbul ignore next
-    if (
-      packageFile.managerData.pnpmShrinkwrap &&
-      config.reuseLockFiles === false
```
Comment (rarkins) [inline lib/modules/manager/npm/post-update/index.ts:232]: yarn also has this logic, I don't think only pnpm should be removed

Comment (viceice) [reply lib/modules/manager/npm/post-update/index.ts:232]: yes, see other PR

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the comments. I'll leave the code review to others.

**Review by secustor — APPROVED** [review]

**Comment by RobHannay** [comment]
You're amazing. Thanks!

**Comment by cristianmadularu** [comment]
Any eta for this? Thanks

**Pull request merged**